### PR TITLE
bigquery/benchmarks: revamp benchmarks

### PR DESCRIPTION
**Pull request by shollyman**
Background:

There's work planned to revisit some of the internal
details about how we run queries, in the effort to improve overall
latency.

Our existing benchmark tool is pretty minimal, and hasn't been used
or touched substantially since 2017.  This revamps that tooling a bit,
and adds some mechanisms for recording benchmark output in bigquery.

Currently the focus is on benchmarking SELECT queries, but in the future
this may be augmented for DML/DDL usages.

### CODE REVIEW

File: bigquery/benchmarks/bench.go [hunk 1]
```diff
@@ -17,69 +17,304 @@
...
+			}
+			return tRef, nil
+		}
+		return nil, fmt.Errorf("error while validating table existence: %v", err)
+	}
+	return tRef, nil
+}
+
+func reportResults(ctx context.Context, client *bigquery.Client, table *bigquery.Table, results []*profiledQuery) error {
+
```
Comment (hongalex) [inline bigquery/benchmarks/bench.go:257]: nit: extra whitespace

File: bigquery/benchmarks/bench.go [hunk 2]
```diff
@@ -17,69 +17,304 @@
...
+	}
+	return nil
+}
+
+func main() {
+	var reruns = flag.Int("reruns", 3, "number of reruns to issue for each query")
+	var queryfile = flag.String("queryfile", "benchmarked-queries.json", "path to file contain queries to be benchmarked.")
+	var projectID = flag.String("projectid", "", "project ID to use for running benchmarks.  Uses PROJECT_ID env if not set.")
+	var reportTable = flag.String("table", "", "table to stream results into, specified in project.dataset.table format")
+	var createTable = flag.Bool("create_table", false, "create result table if it does not exist")
```
Comment (hongalex) [inline bigquery/benchmarks/bench.go:274]: optional: I usually see flags declared outside of `main()` declared like the following. Don't feel too strongly about it though.

```go
var (
        reruns = flag.Int("reruns", 3, "number of reruns to issue for each query")
	queryfile = flag.String("queryfile", "benchmarked-queries.json", "path to file contain queries to be benchmarked.")
        ...
)
```

File: bigquery/benchmarks/bench.go [hunk 3]
```diff
@@ -17,69 +17,304 @@
...
+
+	var tags tags
+	flag.Var(&tags, "tag", "an optional key and value seperated by colon (:) character")
+	flag.Parse()
+
+	// Validate flags.
+	if *reruns <= 0 {
+		log.Fatalf("--reruns should be a positive value")
+	}
+	projID := os.Getenv("PROJECT_ID")
```
Comment (hongalex) [inline bigquery/benchmarks/bench.go:284]: Should this be `GCLOUD_PROJECT` or `GOOGLE_CLOUD_PROJECT` instead? I've seen those used as default env variables.

File: bigquery/benchmarks/bench.go [hunk 4]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2017 Google LLC
+// Copyright 2020 Google LLC
```
[the remark this answers is not in the document]
Comment (shollyman) [reply bigquery/benchmarks/bench.go]: reverted this

File: bigquery/benchmarks/bench.go [hunk 5]
```diff
@@ -17,69 +17,304 @@
...
-			log.Fatal(err)
+			timing.ErrorString = err.Error()
+			return timing
+		}
+		rowCount++
+	}
+	return timing
+}
+
+// runBenchmarks
```
[the remark this answers is not in the document]
Comment (shollyman) [reply bigquery/benchmarks/bench.go]: augmented comment

File: bigquery/benchmarks/bench.go (hunk 2 above)
Comment (shollyman) [reply bigquery/benchmarks/bench.go]: This was done simply to keep them from being globals, as they didn't warrant it.

File: bigquery/benchmarks/bench.go (hunk 3 above)
Comment (shollyman) [reply bigquery/benchmarks/bench.go]: swapped to GOOGLE_CLOUD_PROJECT

**Pull request merged**